### ISSUE TITLE
`scipy.linalg.interpolative.interp_decomp` modifies input matrix 

### ISSUE THREAD

**Issue by mdhaber** [scipy/scipy#9793]
`scipy.linalg.interpolative.interp_decomp` modifies the input matrix when `randomized=False` and input matrix is transposed.

### Reproducing code example:
```
import numpy as np
from scipy.linalg.interpolative import interp_decomp

A = np.array([[-1, -1, -1, 0, 0, 0],
              [0, 0, 0, 1, 1, 1],
              [1, 0, 0, 1, 0, 0],
              [0, 1, 0, 0, 1, 0],
              [0, 0, 1, 0, 0, 1]], dtype=float)

idx, proj = interp_decomp(A.T, rank, rand=False)

print(A)
```
prints
```
[[-1.00000000e+00  1.00000000e+00 -1.00000000e+00 -1.00000000e+00
  -0.00000000e+00 -0.00000000e+00]
 [ 0.00000000e+00  1.73205081e+00 -0.00000000e+00 -5.77350269e-01
  -5.77350269e-01 -5.77350269e-01]
 [-5.77350269e-01  5.77350269e-01  1.15470054e+00 -1.92450090e-01
   3.84900179e-01  3.84900179e-01]
 [-5.77350269e-01  5.77350269e-01 -5.77350269e-01  1.00000000e+00
   8.22780956e-01 -1.58285666e-02]
 [-1.00000000e+00  1.00000000e+00 -1.00000000e+00 -1.00000000e+00
   1.11022302e-16  9.71445147e-17]]
```
which is not the original matrix.

This doesn't occur when `randomized=True` or when the input matrix is `A` instead of `A.T`.

A workaround, of course, is to copy the matrix before passing it in.

### Scipy/Numpy/Python version information:
```
1.3.0.dev0+1b00745 1.15.0 sys.version_info(major=3, minor=6, micro=7, releaselevel='final', serial=0)
```

**Comment by Kai-Striega** [comment]
TLDR: `A` is changed in-place in the FORTRAN subroutine [iddr_id](https://github.com/scipy/scipy/blob/12fe2866a9d489d2e4eabd279d1abb930a0c96b6/scipy/linalg/src/id_dist/src/idd_id.f#L142-L260).  

Looking at the output from the pdb, `A` changes in ``_interpolative_backend.py``:

    > /mnt/c/dev/scipy/scipy/linalg/_interpolative_backend.py(232)iddr_id()
    -> idx, rnorms = _id.iddr_id(A, k)
    (Pdb) display A
    display A: array([[-1.,  0.,  1.,  0.,  0.],
           [-1.,  0.,  0.,  1.,  0.],
           [-1.,  0.,  0.,  0.,  1.],
           [ 0.,  1.,  1.,  0.,  0.],
           [ 0.,  1.,  0.,  1.,  0.],
           [ 0.,  1.,  0.,  0.,  1.]])
    (Pdb) s
    > /mnt/c/dev/scipy/scipy/linalg/_interpolative_backend.py(233)iddr_id()
    -> n = A.shape[1]
    (Pdb) display A
    display A: array([[ 0.        ,  0.        , -0.33333333, -0.33333333, -0.33333333],
           [-0.33333333,  0.        , -0.57735027,  0.78867513, -0.21132487],
           [-0.33333333,  0.        , -0.57735027, -0.21132487,  0.78867513],
           [-0.33333333,  1.        ,  1.        ,  0.        ,  0.        ],
           [-0.        ,  1.        ,  0.        ,  1.        ,  0.        ],
           [-0.        ,  1.        ,  0.        ,  0.        ,  1.        ]])

Which is pretty much a wrapper function of the FORTRAN function (docstring omitted):

https://github.com/scipy/scipy/blob/12fe2866a9d489d2e4eabd279d1abb930a0c96b6/scipy/linalg/_interpolative_backend.py#L231-L235

Line 232 calls the FORTRAN subroutine which stores the value of `proj` in `A`. Then reads the value of `proj` into it's own variable (line 234).

I think the simplest way to *not* modify `A` would be to change line 231 to ``B = np.asfortranarray(A)`` and use `B` throughout the remaining calculation. This may (?) require an extra copy, I'm not sure if this induce a significant performance cost.

@mdhaber The FORTRAN code is well explained but I don't know FORTRAN well enough to apply any changes (I don't really think we should anyway). Could you suggest someone to take a look at that?

**Comment by mdhaber** [comment]
Would that fix the problem? The issue seems to be that 'iddr_id' is changing the array in-place. If you change the name of the variable to B, it will still refer to the array in memory that we're trying to preserve. 
Isn't the issue that 'np.asfortranarray' is not creating a copy? 
This would also make sense because the issue only occurs when we're working on the transpose of a matrix. If it's already in Fortran's column-major order, it doesn't create a copy. 
I think the fix should go in the Python code where you say, but I think you just need to create a copy. That is, use 'ndarray.flags' to check whether the area is already 'F_CONTIGUOUS'. If so, explicitly make a copy.

**Comment by ilayn** [comment]
This check is often done via [Here are many checks ](https://github.com/scipy/scipy/blob/master/scipy/linalg/basic.py#L139) that linalg.solve does. You can start from there and make it more sophisticated if needed. 

The main machinery is 

https://github.com/scipy/scipy/blob/master/scipy/linalg/misc.py#L177-L187

**Comment by Kai-Striega** [comment]
Ok, naively, I thought `np.asfortranarray` always makes a copy. I'll add a test case and see if that fixes it.

**Comment by Kai-Striega** [comment]
I've added some extra checks s.t. ``A = np.asfortranarray(A)`` becomes:

    if A.flags['F_CONTIGUOUS']:
        A = A.copy()
    else:
        A = np.asfortranarray(A)

and no longer changes the user's `A`.

@ilayn @mdhaber Other fortran functions also change the underlying  data  (e.g. ``iddr_svd``, ``iddp_svd``,  ``idd_permuter``). And I suspect a fortran array is not being copied before any of them.

**Comment by mdhaber** [comment]
@kai-striega that's unfortunate, but I think they'll need to be patched individually. We can't have `asfortranarray` making unnecessary copies. Even though some people have assumed it makes copies, I imagine that others have used it correctly. We shouldn't slow down the correct code to fix the incorrect code.
Maybe it would help to add a note about this in the `asfortranarray`  documentation. Even an optional argument `copy=False` that, when true, forces a copy even if the array is already Fortran-style.

**Comment by mdhaber** [comment]
@Kai-Striega BTW this might be a good first PR for someone. It's not bothering me right now, so could we leave it? If we are able to update the documentation (#9951) this would be a good PR for someone to learn on.

**Comment by Kai-Striega** [comment]
Yes, good idea. I've also been drafting an issue for numpy to expose the
'copy' argument. Even if it's not used here, it will be useful to have in
general.

**Comment by Kai-Striega** [comment]
@mdhaber rather than update the documentation we can fix this by changing `np.asfortranarray(...) ` to `np.array(a, dtype, copy=True, order='F', ndmin=1)`. But if these are changed it should happen for *all* related fortran functions where the values need to be copied.
